### PR TITLE
Update server.js to trust UserAccounts securely

### PR DESCRIPTION
**Pull request by steve02081504**
we don't need basic auth when we have enableAccounts

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Cohee1207** [comment]
I don't agree with this, since user account passwords are optional. You can dismiss basic auth only if all enabled accounts are password protected.

**Comment by steve02081504** [comment]
> I don't agree with this, since user account passwords are optional. You can dismiss basic auth only if all enabled accounts are password protected.

how about [`585c75c` (#2447)](https://github.com/SillyTavern/SillyTavern/pull/2447/commits/585c75c608af6739f7e0bd69a79aa7d5bfdfa01e)?

**Comment by Cohee1207** [comment]
Slightly better.

@Wolfsblvt thought?

**Comment by Wolfsblvt** [comment]
Don't really know how I'd test that.

You moved logging the unprotected password accounts to `postSetupTasks`, but that won't run in your case if its not set up correctly and missing passwords, as you exit out before.
Doing the logging only once makes sense, but it should be before exiting out.

Otherwise I think it's sound to argue starting unprotected but with only secured accounts is okay.  
I feel like we should still log a warning. Or still require the security override switch. Because... even if you have protected accounts, you should be warned that your ST instance is now public and accessible to everyone. There is no auth or protection in front of it anymore, its just locked behind user accounts.

**Comment by steve02081504** [comment]
> I feel like we should still log a warning

more logs added in [`c85a430` (#2447)](https://github.com/SillyTavern/SillyTavern/pull/2447/commits/c85a430fb25f30086c9b3b7499522c2470aab5f3), any other idea?

File: server.js [hunk 1]
```diff
@@ -658,6 +645,33 @@ function setWindowTitle(title) {
...
+                console.warn(color.red('Security has been overridden. If it\'s not a trusted network, change the settings.'));
+                return
+            }
+            if(!enableWhitelist && !basicAuthMode && !enableAccounts) {
+                console.error(color.red('Your SillyTavern is currently unsecurely open to the public. Enable whitelisting or basic authentication or accounts authentication.'));
+                process.exit(1);
+            }
+            const users = await userModule.getAllUsers();
+            const unprotectedUsers = users.filter(x => x.admin && !x.password);
+            if (unprotectedUsers.length > 0) {
```
Comment (Cohee1207) [inline server.js]: I think it should only end the process if the basic auth mode is disabled. Reporting it otherwise is fine.

Comment (steve02081504) [reply server.js]: fixed in [`422306c` (#2447)](https://github.com/SillyTavern/SillyTavern/pull/2447/commits/422306c2b6faed71601eb74e42f54c20294cc328)

**Comment by Cohee1207** [comment]
Consider the following table (co-authored with Claude 3.5). 
@steve02081504 @Wolfsblvt how much sense does it make?

| Basic auth | Whitelist | Listen | Enabled accounts | Unprotected accounts | Outcome |
| :--------- | :-------- | :----- | :--------------- | :------------------- | :------ |
| False      | False     | False  | False            | False                | Allow   |
| False      | False     | True   | False            | False                | Exit    |
| True       | True      | True   | True             | True                 | Warn    |
| True       | False     | True   | True             | True                 | Exit    |
| False      | True      | True   | True             | True                 | Warn    |
| True       | True      | True   | False            | False                | Allow   |
| False      | False     | True   | True             | False                | Allow   |

**Comment by Wolfsblvt** [comment]
That's a bit confusing, yeah. Hmmh...

My idea:
1. **listen** false is easy, that's always just allow
2. If you have either **basic auth** or **whitelist** set, it will never exit, just warn or allow
3. Unprotected accounts (with listen on) will always log a warning (no matter if admin or not)
4. If both **basic auth** and **whitelist** are off, having unprotected admin accounts will exit, while not will log an additional warning that just having accounts is "not the best protection" or whatever.

**Comment by steve02081504** [comment]
I tried to improve the code a little, but I don't think it will please everyone.
If there's anything you find wrong or want to change. Feel free to👍.

**Comment by Cohee1207** [comment]
> I think it's only necessary to exit the program when there's unprotected admin users

It's not good to facilitate unauthorized access to any account, admin or not. But let's settle on that for now.

**Pull request merged**